src/sync/deduplicator.ts

```diff
@@ -0,0 +1,200 @@
+		const REDIS_CLEANUP_TIMEOUT = 24 * 3600 * 1000;
+		await this.redis.set(jobKey, JSON.stringify(flag), 'px', REDIS_CLEANUP_TIMEOUT);
+	}
+
+	async isExecutionLive(jobKey: string, executionId: string, executorFlagUpdateTimeoutMsecs: number): Promise<boolean> {
+		const flagOld = await this.redis.get(jobKey);
+		if (!flagOld) {
+			// Cannot tell, might have finished already (race condition)
+			return false;
+		}
+		const flagOldParsed = JSON.parse(flagOld) as Flag;
+		if (flagOldParsed.executionId !== executionId) {
+			// Cannot tell, other node might haved pick up the job and updated the flag
+			return false;
+		}
+
+		await sleep(executorFlagUpdateTimeoutMsecs * 2);
+
+		const flagNew = await this.redis.get(jobKey);
+		if (!flagNew) {
+			// Cannot tell, might have been deleted by Redis during auto-expiration
+			return false;
+		}
+		const flagNewParsed = JSON.parse(flagNew) as Flag;
+		// The only condition when we can say for sure that something is processing the job
+		return flagNewParsed.executionId === executionId && flagOldParsed.timestamp < flagNewParsed.timestamp;
+	}
+}
+
+export enum DeduplicatorResult {
+	E_OK,
+	E_OTHER_WORKER_DOING_THIS_JOB,
+	E_NOT_SURE_TRY_AGAIN_LATER
+}
+
+export class Deduplicator {
+	private inProgressStorage: InProgressStorage;
+	private executorFlagUpdateTimeoutMsecs: number;
+
+	constructor(inProgressStorage: InProgressStorage, executorFlagUpdateTimeoutMsecs: number) {
+		this.inProgressStorage = inProgressStorage;
+		this.executorFlagUpdateTimeoutMsecs = executorFlagUpdateTimeoutMsecs;
+	}
+
+	/**
+	 * Executes the job if and only if it can determine that there's no other executor running the job (by checking the
+	 * flag in the storage). If it can't tell and there's a flag, tries to understand if that other executor is live or
+	 * this is some weird state.
+	 *
+	 * During the execution, refreshes the flag each executorFlagUpdateTimeoutMsecs to keep other executors informed.
+	 *
+	 * Based on the outcome of the call, you might decide to do the following:
+	 * 	E_OK - the job has been processed, the flag was dropped, ANOTHER JOB with the same key can be scheduled and
+	 * 				 it WILL BE PROCESSED NEXT
+	 *  E_OTHER_WORKER_DOING_THIS_JOB - we have determined that the current task has been processed right now,
+	 *  																you might want to DISCARD the job to avoid duplication
+	 *  E_UNSURE_TRY_AGAIN_LATER - we are not sure if the other process is processing the job right now or not; you might like to
+	 *  					  						   RESCHEDULE the task for good (remember there's a timeout for the flag)
+	 *  <an exception is thrown> - same as E_OK with the only difference that the underlying job threw an exception; the
+	 *  													 flag was removed.
+	 */
+	async executeWithDeduplication(jobKey: string, jobRunner: () => Promise<void>): Promise<DeduplicatorResult> {
+		const executionId = 'execid' + Math.random();
```

Comment:
Let's use `uuid` here. It is a de facto standard for random identifiers generation around the code

Also we need to add this executionId to logger parameters. Also we need to choose different name for it, hence we currently have "executionId" parameter in logs added by SQS Listener, which is message processing attempt ID

Reply:
- [ ] Use uuid instead of number
- [ ] Rename "executionId" to something else

Reply:
Also we can solve renaming problem if we take executionID from the SQS processing context. So we'll use the same identifier for SQS message processing attempt ID and for deduplication.